Python 3.12 - ValueError: generator already executing

### Bug description

It seems that with a combination of a commit to astroid, and a change in Python 3.12, I am getting a `ValueError: generator already executing`.

* https://github.com/pylint-dev/astroid/commit/c5352d53bc728816836a217428cac2eb305a0d55
* https://github.com/python/cpython/commit/6968f9e4d3593610b60c1140f04de275ff40cd41

Unfortunately, despite about 12 hours of trying, I don't have a simple reproducer, and this will involve running pylint against https://github.com/ansible/ansible

I've logged here instead of directly against astroid, as I was not able to find an easy way to cause the error with astroid alone.


### Configuration

```ini
N/A
```


### Command used

```shell
git clone git@github.com:ansible/ansible.git
cd ansible
python3.12 -m pylint lib/ansible/playbook/base.py > /dev/null
```


### Pylint output

```python
Traceback (most recent call last):
  File ".../astroid/astroid/decorators.py", line 104, in inner
    yield from generator
ValueError: generator already executing
```

If I insert additional debugging into astroid, to use `traceback.print_stack()` when that `ValueError` happens, I see something like:

<details>
<summary>Traceback...</summary>
<p>

```
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File ".../lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File ".../lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File ".../lib/python3.12/site-packages/pylint/lint/run.py", line 211, in __init__
    linter.check(args)
  File ".../lib/python3.12/site-packages/pylint/lint/pylinter.py", line 704, in check
    self._lint_files(ast_per_fileitem, check_astroid_module)
  File ".../lib/python3.12/site-packages/pylint/lint/pylinter.py", line 752, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File ".../lib/python3.12/site-packages/pylint/lint/pylinter.py", line 788, in _lint_file
    check_astroid_module(module)
  File ".../lib/python3.12/site-packages/pylint/lint/pylinter.py", line 1017, in check_astroid_module
    retval = self._check_astroid_module(
  File ".../lib/python3.12/site-packages/pylint/lint/pylinter.py", line 1069, in _check_astroid_module
    walker.walk(node)
  File ".../lib/python3.12/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File ".../lib/python3.12/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File ".../lib/python3.12/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File ".../lib/python3.12/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File ".../lib/python3.12/site-packages/pylint/checkers/typecheck.py", line 1440, in visit_call
    called, implicit_args, callable_name = _determine_callable(called)
  File ".../lib/python3.12/site-packages/pylint/checkers/typecheck.py", line 636, in _determine_callable
    new = callable_obj.local_attr("__new__")[-1]
  File ".../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2327, in local_attr
    class_node = next(self.local_attr_ancestors(name, context), None)
  File ".../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2279, in local_attr_ancestors
    ancestors: Iterable[ClassDef] = self.mro(context)[1:]
  File ".../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2924, in mro
    return self._compute_mro(context=context)
  File ".../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2893, in _compute_mro
    inferred_bases = list(self._inferred_bases(context=context))
  File ".../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2873, in _inferred_bases
    baseobj = next(
  File ".../astroid/astroid/decorators.py", line 56, in wrapped
    yield res
  File ".../astroid/astroid/bases.py", line 181, in _infer_stmts
    yield inf
  File ".../astroid/astroid/nodes/node_ng.py", line 175, in infer
    yield result
  File ".../astroid/astroid/decorators.py", line 56, in wrapped
    yield res
  File ".../astroid/astroid/bases.py", line 181, in _infer_stmts
    yield inf
  File ".../astroid/astroid/nodes/node_ng.py", line 175, in infer
    yield result
  File ".../astroid/astroid/decorators.py", line 56, in wrapped
    yield res
  File ".../astroid/astroid/bases.py", line 181, in _infer_stmts
    yield inf
  File ".../astroid/astroid/nodes/node_ng.py", line 175, in infer
    yield result
  File ".../astroid/astroid/decorators.py", line 56, in wrapped
    yield res
  File ".../astroid/astroid/nodes/node_ng.py", line 175, in infer
    yield result
  File ".../astroid/astroid/decorators.py", line 107, in inner
    traceback.print_stack(file=sys.stderr)
```

</p>
</details>


### Expected behavior

No traceback

### Pylint version

```shell
astroid 3.0.0
pylint 3.0.1
Python 3.12.0
```


### OS / Environment

N/A

### Additional dependencies

```
N/A
```

The exception output is interspersed in the normal output from pylint.  So it doesn't seem necessarily to happen at the very end.  I'm unsure exactly the internal timing of things, it could be in object finalization, but it doesn't seem to be after the main program finishes.

```
[SNIP]
lib/ansible/playbook/base.py:435:12: W0201: Attribute '_ds' defined outside __init__ (attribute-defined-outside-init)
lib/ansible/playbook/base.py:716:0: C0115: Missing class docstring (missing-class-docstring)
Traceback (most recent call last):
  File "/Users/sivel/projects/ansibledev/astroid/astroid/decorators.py", line 104, in inner
    yield from generator
ValueError: generator already executing
lib/ansible/playbook/base.py:739:58: E1101: Module 'ansible.constants' has no 'ANY_ERRORS_FATAL' member (no-member)
lib/ansible/playbook/base.py:741:48: E1101: Module 'ansible.constants' has no 'TASK_TIMEOUT' member (no-member)
[SNIP]
```

--- a/.pyenchant_pylint_custom_dict.txt
+++ b/.pyenchant_pylint_custom_dict.txt
@@ -361,6 +361,7 @@ unicode
 Uninferable
 uninferable
 unittest
+unraisablehook
 untriggered
 # prefix for string
 ur

--- a/doc/whatsnew/fragments/9138.bugfix
+++ b/doc/whatsnew/fragments/9138.bugfix
@@ -1,0 +1,4 @@
+Catch incorrect ValueError ``"generator already executing"`` for Python 3.12.0 - 3.12.2.
+This is fixed upstream in Python 3.12.3.
+
+Closes #9138

--- a/pylint/__init__.py
+++ b/pylint/__init__.py
@@ -95,4 +95,25 @@ def modify_sys_path() -> None:
         sys.path.pop(1)
 
 
+def _catch_valueerror(unraisable: sys.UnraisableHookArgs) -> None:
+    """Overwrite sys.unraisablehook to catch incorrect ValueError.
+
+    Python 3.12 introduced changes that sometimes cause astroid to emit ValueErrors
+    with 'generator already executing'. Fixed in Python 3.12.3 and 3.13.
+
+    https://github.com/pylint-dev/pylint/issues/9138
+    """
+    if (
+        isinstance(unraisable.exc_value, ValueError)
+        and unraisable.exc_value.args[0] == "generator already executing"
+    ):
+        return
+
+    sys.__unraisablehook__(unraisable)
+
+
+if (3, 12, 0) <= sys.version_info[:3] < (3, 12, 3):
+    sys.unraisablehook = _catch_valueerror
+
+
 version = __version__

--- a/pylint/__init__.py
+++ b/pylint/__init__.py
@@ -95,7 +95,7 @@ def modify_sys_path() -> None:
         sys.path.pop(1)
 
 
-def _catch_valueerror(unraisable: sys.UnraisableHookArgs) -> None:
+def _catch_valueerror(unraisable: sys.UnraisableHookArgs) -> None:  # pragma: no cover
     """Overwrite sys.unraisablehook to catch incorrect ValueError.
 
     Python 3.12 introduced changes that sometimes cause astroid to emit ValueErrors

